File: Algorithms_python/56_Merge Intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda x: x[0])
        st = ed = -1
        res = []
        for s, e in intervals:
            if ed < s:
                if st != -1:
                    res.append([st, ed])
                st, ed = s, e
            else:
                ed = max(ed, e)
        if st != -1:
            res.append([st, ed])
        return res
'''
创一个上线边界，在循环list的时候挨个比较，下一个数组越过了上一个的上边界就存储，
没越过就扩展边界，在下一次循环储存
方法有两个注意点，首先要整理数组，要从小到大排列
其次是最后一次会没保存，要额外存储一下
'''

File: Algorithms_C#/56_Merge_Intervals.cs
public class Solution {
    public int[][] Merge(int[][] intervals) {
        intervals = intervals.OrderBy(a => a[0]).ToArray();
        IList<int[]> ans = new List<int[]>();
        ans.Add(intervals[0]);
        for (int i = 0; i < intervals.Length; ++i){
            if (ans[ans.Count - 1][1] < intervals[i][0]){
                ans.Add(intervals[i]);
            }
            else{
                ans[ans.Count - 1][1] = Math.Max(ans[ans.Count - 1][1], intervals[i][1]);
            }
        }
        return ans.ToArray();
    }
}